STM32F1/libraries/EEPROM/EEPROM.cpp

```diff
@@ -212,56 +212,71 @@ uint16 EEPROMClass::EE_VerifyPageFullWriteVariable(uint16 Address, uint16 Data)
 	FLASH_Status FlashStatus;
 	uint32 idx, pageBase, pageEnd, newPage;
 	uint16 count;
-
+	boolean found = false;
 	// Get valid Page for write operation
+	
 	pageBase = EE_FindValidPage();
 	if (pageBase == 0)
 		return  EEPROM_NO_VALID_PAGE;
-
+	
 	// Get the valid Page end Address
 	pageEnd = pageBase + PageSize;			// Set end of page
-
-	for (idx = pageEnd - 2; idx > pageBase; idx -= 4)
+	
+	//FLASH_Unlock();
+	
+	for (idx = pageEnd - 2; idx >= pageBase; idx -= 4)
 	{
 		if ((*(__IO uint16*)idx) == Address)		// Find last value for address
 		{
+			found = false;
+			
 			count = (*(__IO uint16*)(idx - 2));	// Read last data
+			
 			if (count == Data)
 				return EEPROM_OK;
-			if (count == 0xFFFF)
+			
+			if (count == 0xFFFF || Data < count)
 			{
-				FlashStatus = FLASH_ProgramHalfWord(idx - 2, Data);	// Set variable data
-				if (FlashStatus == FLASH_COMPLETE)
-					return EEPROM_OK;
+			
+			FlashStatus = FLASH_ProgramHalfWord(idx - 2, Data);	// Set variable data
+			if (FlashStatus == FLASH_COMPLETE)
+				return EEPROM_OK;
 			}
+			
 			break;
+		} else {
+			found = true;
+			
 		}
 	}
-
-	// Check each active page address starting from begining
-	for (idx = pageBase + 4; idx < pageEnd; idx += 4)
-		if ((*(__IO uint32*)idx) == 0xFFFFFFFF)				// Verify if element 
-		{													//  contents are 0xFFFFFFFF
-			FlashStatus = FLASH_ProgramHalfWord(idx, Data);	// Set variable data
-			if (FlashStatus != FLASH_COMPLETE)
-				return FlashStatus;
-			FlashStatus = FLASH_ProgramHalfWord(idx + 2, Address);	// Set variable virtual address
-			if (FlashStatus != FLASH_COMPLETE)
-				return FlashStatus;
-			return EEPROM_OK;
+	if(found){
+		// Check each active page address starting from begining
+		for (idx = pageBase + 4; idx < pageEnd; idx += 4){
+			if ((*(__IO uint32*)idx) == 0xFFFFFFFF)				// Verify if element 
+			{													//  contents are 0xFFFFFFFF
+				FlashStatus = FLASH_ProgramHalfWord(idx, Data);	// Set variable data
+				if (FlashStatus != FLASH_COMPLETE)
+					return FlashStatus;
+					FlashStatus = FLASH_ProgramHalfWord(idx + 2, Address);	// Set variable virtual address
+					if (FlashStatus != FLASH_COMPLETE)
+						return FlashStatus;
+					return EEPROM_OK;
```

Comment:
I think the indentation of these lines is not correct, please check and correct.